crates/runtime/src/component/mod.rs

```diff
@@ -192,9 +198,97 @@ where
     experimental_features: Features,
 }
 
+/// The [`CustomCtxComponent`] is similar to [`Component`], but it supports passing a custom context that
+/// implements the [`BaseCtx`] trait. This is useful for when you want to extend the context with additional
+/// host implementations or custom functionality.
+#[derive(Clone)]
+pub struct CustomCtxComponent<C>
+where
+    C: BaseCtx,
+{
+    engine: wasmtime::Engine,
+    instance_pre: wasmtime::component::InstancePre<C>,
+    #[allow(unused)]
+    host_resources: Arc<HashMap<Box<str>, HashMap<Box<str>, (ResourceType, ResourceType)>>>,
+    max_execution_time: Duration,
+}
+
+impl<C> CustomCtxComponent<C>
+where
+    C: BaseCtx,
+{
+    /// Compiles a WebAssembly component using [Runtime]. Modules are not supported with this function.
+    ///
+    /// It is expected that the `linker_fn` is used to link any and all imports to the component. If
+    /// not, then the instantiation of the instance will fail.
+    #[instrument(level = "trace", skip_all)]
+    pub fn new_with_linker_minimal(
+        rt: &Runtime,
+        wasm: &[u8],
+        linker_fn: impl FnOnce(&mut Linker<C>, &wasmtime::component::Component) -> anyhow::Result<()>,
+    ) -> anyhow::Result<Self> {
+        if wasmparser::Parser::is_core_wasm(wasm) {
+            anyhow::bail!("core modules are not supported in the minimal linker");
+        }
+        let engine = rt.engine.clone();
+        let component = wasmtime::component::Component::new(&engine, wasm)
+            .context("failed to compile component")?;
+
+        let mut linker = Linker::new(&engine);
+
+        let ty = component.component_type();
+        let mut guest_resources = Vec::new();
+        let mut host_resources = BTreeMap::new();
```

Comment:
Super duper nit: Any reason for the BTreeMap here over HashMap? Doesn't really matter in this case unless the ordering matters but a HashMap generally has better iteration performance

Reply:
Nope, fully copied from elsewhere in the code 😆